### DISCUSSION TITLE
Web Scrapping - looping untuk column yang berbeda namun memiliki HMTL format sama

### DISCUSSION THREAD

**Discussion by fazahanifandra** [teamalgoritma/community#908] [Q&A] [answered]
Selamat pagi bapak & ibu, saya ingin bertanya cara melakukan looping untuk web scrapping dimana columnnya yang memiliki format html yang sama.

Saya sudah membaca di https://github.com/teamalgoritma/community/discussions/670 untuk cara mengambil informasi dari format html yang sama. Berikut screenshot kendala awalnya dimana column: 'Worldwide', 'Domestic' dan 'Foreign' memiliki format yang sama (class= 'a-text-right mojo-field-type-money'):

![image](https://github.com/user-attachments/assets/542e229b-d001-43ee-a86c-c5f1187e8d0e) 
![image](https://github.com/user-attachments/assets/a90287d2-951c-4e05-ab54-8dedd4b0ea6c)

Berikut potongan kode saya:

1. Di awal saya sudah berhasil finding the right key:
```
table = soup.find('div', attrs={'class':'a-section imdb-scroll-table-inner'})
print(table.prettify()[1:500])
```
2. kemudian saya ikutin petunjuk https://github.com/teamalgoritma/community/discussions/670, untuk mengambil informasi dari 3 item berbeda jika memiliki format HTML yang sama:

```
money_elements = table.find_all('td', attrs='a-text-right mojo-field-type-money')
if len(money_elements) >= 3:
    worldwide = money_elements[0].text.strip()
    domestic = money_elements[1].text.strip()
    foreign = money_elements[2].text.strip()
    print(worldwide)
    print(domestic)
    print(foreign)
```
3. Namun saya kurang paham dari kode atas tersebut, cara memasukan ke looping for scrapping agar bisa keluar nilai masing-masing 3 column tersebut untuk 200 rows datanya:

```
temp = []  # Initiating a tuple

for i in range(0, row_length):

    # Scraping process
    # Get Rank
    rank = table.find_all('td', attrs='a-text-right mojo-header-column mojo-truncate mojo-field-type-rank mojo-sort-column')[i].text

    # Get Release Group
    release_group = table.find_all('td', attrs='a-text-left mojo-field-type-release_group')[i].text

    # Get Worldwide
    worldwide = money_elements[i].text

    # Get Domestic
    domestic = money_elements[i].text

    # Get Foreign
    foreign = money_elements[i].text

    temp.append((rank, release_group, worldwide, domestic, foreign))
    
temp
```
Saya sudah berhasil untuk extract Rank & Release Group, namun kalau seperti ini akan memberikan angka / data yang sama untuk kolom 'worldwide', 'domestic' dan 'foreign'. Mohon bantuannya dan terima kasih banyak

**Comment by saskia-dwi-ulfah** [comment]
Dear @fazahanifandra 

Terima kasih telah bertanya melalui Github Discussion Algoritma.
Apakah bisa dibantu untuk melampirkan keseluruhan kodenya, pak? Terutama di bagian awal mulai scraping hingga proses looping yang Bapak lakukan. Hal ini akan membantu kami untuk tracing dan solving kendala yang Bapak alami.

Best regards,
Team Algoritma

**Reply by fazahanifandra** [reply]
Baik terima kasih atas responnya. Berikut saya lampirkan keseluruhan kodenya:

```
import requests
url_get = requests.get('https://www.boxofficemojo.com/year/world/')
```
`url_get.content[1:500]`

```
from bs4 import BeautifulSoup 

soup = BeautifulSoup(url_get.content,"html.parser")
print(type(soup))
```
```
table = soup.find('div', attrs={'class':'a-section imdb-scroll-table-inner'})
print(table.prettify()[1:500])
```
```
table_all = table.find_all('td', attrs={'class','a-text-right mojo-header-column mojo-truncate mojo-field-type-rank mojo-sort-column'})
table_all
```
```
row_length = len(table_all)
row_length
```
```
money_elements = table.find_all('td', attrs='a-text-right mojo-field-type-money')
if len(money_elements) >= 3:
    worldwide = money_elements[0].text.strip()
    domestic = money_elements[1].text.strip()
    foreign = money_elements[2].text.strip()
    print(worldwide)
    print(domestic)
    print(foreign)
```
```
temp = []  # Initiating a tuple

for i in range(0, row_length):

    # Scraping process
    # Get Rank
    rank = table.find_all('td', attrs='a-text-right mojo-header-column mojo-truncate mojo-field-type-rank mojo-sort-column')[i].text.strip()

    # Get Release Group
    release_group = table.find_all('td', attrs='a-text-left mojo-field-type-release_group')[i].text.strip()

    # Get Worldwide
    worldwide = money_elements[i].text.strip().replace('$', '')

    # Get Domestic
    domestic = money_elements[i].text.strip().replace('$', '')

    # Get Foreign
    foreign = money_elements[i].text.strip().replace('$', '')
    
    temp.append((rank, release_group, worldwide, domestic, foreign))
    
temp
```

**Comment by saskia-dwi-ulfah** [comment, marked as answer]
Dear @fazahanifandra 

Setelah kami coba kode yang Anda lampirkan, metrics Worldwide, Domestic, dan Foreign tersimpan pada `money_elements`. Karena ada 200 data, maka pada `money_elements` terdapat 600 data. Di mana: 3 element pertama merupakan data untuk baris pertama, 3 element kedua merupakan data untuk baris kedua, dst.

Oleh karena itu, terdapat sedikit modifikasi bagaimana kita mengambil nilai2 tersebut dari `money_elements`. Berikut kami lampirkan:
```
temp = []  # Initiating a tuple

for i in range(0, row_length):

    # Scraping process
    # Get Rank
    rank = table.find_all('td', attrs='a-text-right mojo-header-column mojo-truncate mojo-field-type-rank mojo-sort-column')[i].text.strip()

    # Get Release Group
    release_group = table.find_all('td', attrs='a-text-left mojo-field-type-release_group')[i].text.strip()

    worldwide = money_elements[i*3].text.strip().replace('$', '')
    domestic = money_elements[i*3 + 1].text.strip().replace('$', '')
    foreign = money_elements[i*3 + 2].text.strip().replace('$', '')
    
    temp.append((rank, release_group, worldwide, domestic, foreign))
    
temp
```

di mana dengan kode di atas:

* Kita mengambil data pada indeks kelipatan 3 dari `money_elements` untuk nilai worldwide setiap barisnya.
* Kita mengambil data pada indeks kelipatan 3 + 1 dari `money_elements` untuk nilai domestic setiap barisnya.
* Kita mengambil data pada indeks kelipatan 3 + 2 dari `money_elements` untuk nilai foreign setiap barisnya.


Semoga jawaban ini dapat membantu.

Best regards,
Team Algoritma

**Reply by fazahanifandra** [reply]
Baik terima kasih banyak atas bantuannya, sudah berhasil kode tersebut tanpa errror :)
